airflow/api_fastapi/core_api/services/ui/grid.py

```diff
@@ -0,0 +1,269 @@
+        else (
+            node.get_mapped_ti_count(run_id=run_id, session=session)
+            if isinstance(node, (MappedTaskGroup, MappedOperator))
+            else node
+        )
+        for node in task_count
+    )
+
+
+def fill_task_instance_summaries(
+    grouped_task_instances: dict[tuple[str, str], list],
+    task_instance_summaries_to_fill: dict[str, list],
+    task_node_map: dict[str, dict[str, Any]],
+    session: SessionDep,
+) -> None:
+    """
+    Fill the Task Instance Summaries for the Grouped Task Instances.
+
+    :param grouped_task_instances: Grouped Task Instances
+    :param task_instance_summaries_to_fill: Task Instance Summaries to fill
+    :param task_node_map: Task Node Map
+    :param session: Session
+
+    :return: None
+    """
+    ## Additional logic to calculate the overall state and task count dict of states
+    priority: list[None | TaskInstanceState] = [
+        TaskInstanceState.FAILED,
+        TaskInstanceState.UPSTREAM_FAILED,
+        TaskInstanceState.UP_FOR_RETRY,
+        TaskInstanceState.UP_FOR_RESCHEDULE,
+        TaskInstanceState.QUEUED,
+        TaskInstanceState.SCHEDULED,
+        TaskInstanceState.DEFERRED,
+        TaskInstanceState.RUNNING,
+        TaskInstanceState.RESTARTING,
+        None,
+        TaskInstanceState.SUCCESS,
+        TaskInstanceState.SKIPPED,
+        TaskInstanceState.REMOVED,
+    ]
+
+    overall_states: dict[tuple[str, str], str] = {
+        (task_id, run_id): next(
+            (str(state.value) for state in priority for ti in tis if state is not None and ti.state == state),
+            "no_status",
+        )
+        for (task_id, run_id), tis in grouped_task_instances.items()
+    }
+    for (task_id, run_id), tis in grouped_task_instances.items():
+        ti_try_number = max([ti.try_number for ti in tis])
+        ti_start_date = min([ti.start_date for ti in tis if ti.start_date], default=None)
+        ti_end_date = max([ti.end_date for ti in tis if ti.end_date], default=None)
+        ti_queued_dttm = min([ti.queued_dttm for ti in tis if ti.queued_dttm], default=None)
+        ti_note = min([ti.note for ti in tis if ti.note], default=None)
+
+        all_states = {"no_status" if state is None else state.name.lower(): 0 for state in priority}
+        # Update Task States for non-grouped tasks
+        all_states.update(
+            {
+                "no_status" if state is None else state.name.lower(): len(
+                    [ti for ti in tis if ti.state == state]
+                    if not task_node_map[task_id]["is_group"]
+                    else [
+                        ti
+                        for ti in tis
+                        if ti.state == state and ti.task_id in get_child_task_map(task_id, task_node_map)
+                    ]
+                )
+                for state in priority
+            }
+        )
+        # Update Nested Task Group
+        all_states.update(
+            {
```

Comment:
Why is all that not similar to the legacy implementation ? What are we trying to achieve by re-implementing that ?

Reply:
What do you mean by all? The service as a whole? 
I think populating the task mapping in different styles and splitting methods into smaller pieces such as filling the parents and children in iteration rather than in a single method cause most of the changes.